pkg/controller/clusteroperation/clusteroperation.go

```diff
@@ -0,0 +1,158 @@
+Modifications Copyright 2019 The Libopenstorage Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusteroperation
+
+import (
+	"context"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"reflect"
+	"time"
+
+	corev1alpha1 "github.com/libopenstorage/operator/pkg/apis/core/v1alpha1"
+	"github.com/portworx/sched-ops/k8s"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// ControllerName is the name of the controller
+	ControllerName      = "clusteroperation-controller"
+	validateCRDInterval = 5 * time.Second
+	validateCRDTimeout  = 1 * time.Minute
+)
+
+type Controller struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+func (c *Controller) RegisterCRD() error {
+	return c.createCRD()
+}
+
+func (c *Controller) Init(mgr manager.Manager) error {
+
+	c.client = mgr.GetClient()
+
+	ctrl, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: c})
+	if err != nil {
+		return err
+	}
+
+	// Create a source for watching ClusterOperation events.
+	src := &source.Kind{Type: &corev1alpha1.ClusterOperation{}}
+	// Create a handler for handling events from ClusterOperation resource.
+	h := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &corev1alpha1.ClusterOperation{},
+	}
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			logrus.Info("CREATE")
+			err := c.updateClusterOperationStatus(e.Object)
+			if err != nil {
+				//TODO: handle
+			}
+			//TODO: sdk call
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			logrus.Info("UPDATE")
+			//TODO: sdk call
+			return e.MetaOld.GetGeneration() != e.MetaNew.GetGeneration()
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			logrus.Info("DELETE")
```

Comment:
instead of using logrus directly, add a log util for the cluster operation object. See https://github.com/libopenstorage/stork/blob/master/pkg/log/log.go as reference.

Reply:
logrus provides a convenient formatter, that allows date/time/path_to_file/line in logs. I'm planning to modify it soon